apis/v1beta1/config.go

```diff
@@ -445,67 +447,133 @@ const (
+		}
 
-	port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
-	if err != nil {
-		errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
-			telemetry.Metrics.Address)
-		logger.Info(errMsg, "error", err)
-		return "", 0, err
-	}
+		port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
+		if err != nil {
+			errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
+				telemetry.Metrics.Address)
+			logger.Info(errMsg, "error", err)
+			return "", 0, err
+		}
 
-	host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
-	return host, int32(port), nil
+		host, _, _ := strings.Cut(telemetry.Metrics.Address, explicitPortMatches[0])
+		return host, intToInt32Safe(int(port)), nil
+	}
+
+	port := defaultServicePort
+	host := defaultServiceHost
+	for _, reader := range telemetry.Metrics.Readers {
+		if reader.Pull != nil {
+			if reader.Pull.Exporter.Prometheus != nil {
+				if reader.Pull.Exporter.Prometheus.Host != nil && *reader.Pull.Exporter.Prometheus.Host != "" {
+					host = *reader.Pull.Exporter.Prometheus.Host
+				}
+				if reader.Pull.Exporter.Prometheus.Port != nil && *reader.Pull.Exporter.Prometheus.Port != 0 {
+					port = intToInt32Safe(int(*reader.Pull.Exporter.Prometheus.Port))
+				}
+				break
+			}
+		}
+	}
+	return host, port, nil
 }
 
 // ApplyDefaults inserts configuration defaults if it has not been set.
 func (s *Service) ApplyDefaults(logger logr.Logger) error {
-	telemetryAddr, telemetryPort, err := s.MetricsEndpoint(logger)
-	if err != nil {
-		return err
+	tel := s.GetTelemetry()
+
+	if tel == nil {
+		tel = &Telemetry{}
 	}
+	s.configureMetricReaders(logger, tel)
 
-	tm := &AnyConfig{
-		Object: map[string]interface{}{
-			"metrics": map[string]interface{}{
-				"address": fmt.Sprintf("%s:%d", telemetryAddr, telemetryPort),
-			},
-		},
+	// This field is deprecated and should not be used anymore.
+	if tel.Metrics.Address != "" {
+		// Setting the "address" field to an empty string explicitly removes the value
+		// during Kubernetes serialization. Directly deleting the field from the map using
+		// delete(metrics, "address") does not work because Kubernetes treats missing fields
+		// differently from explicitly empty ones. By assigning "", we ensure the configuration
+		// is updated correctly when the resource is persisted.
+		tel.Metrics.Address = ""
 	}
 
-	if s.Telemetry == nil {
-		s.Telemetry = tm
-		return nil
+	yamlConfig, err := tel.toMap()
+	if err != nil {
+		return err
 	}
-	// NOTE: Merge without overwrite. If a telemetry endpoint is specified, the defaulting
-	// respects the configuration and returns an equal value.
-	if err := mergo.Merge(s.Telemetry, tm); err != nil {
-		return fmt.Errorf("telemetry config merge failed: %w", err)
+
+	s.Telemetry = &AnyConfig{
+		Object: yamlConfig,
 	}
+
 	return nil
 }
 
+func (s *Service) configureMetricReaders(logger logr.Logger, tel *Telemetry) {
```

Comment:
suggestion: can we make this function idempotent? Should be easier to test this and then also test in the future when we expand our own capabilities.

Reply:
also we could just return a raw map here which would mean we could remove the toMap function. I don't _love_ using a map here instead of real types, but i'd rather this than needing that marshal/unmarshal below. @swiatekm thoughts?